File: 3225-length-of-longest-subarray-with-at-most-k-frequency/length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums, k):
        freq = {}
        left = 0
        max_len = 0

        for right, x in enumerate(nums):
            freq[x] = freq.get(x, 0) + 1

            while freq[x] > k:
                y = nums[left]
                freq[y] -= 1
                left += 1

            length = right - left + 1
            if length > max_len:
                max_len = length

        return max_len

File: 3225-length-of-longest-subarray-with-at-most-k-frequency/length-of-longest-subarray-with-at-most-k-frequency.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    int maxSubarrayLength(std::vector<int>& nums, int k) {
        std::unordered_map<int, int> freq;
        int left = 0, right = 0, max_length = 0;
        while (right < nums.size()) {
            freq[nums[right]]++;
            while (freq[nums[right]] > k) {
                freq[nums[left]]--;
                left++;
            }
            max_length = std::max(max_length, right - left + 1);
            right++;
        }
        return max_length;
    }
};